README.md
# checkbox_grouped
![pub](https://img.shields.io/badge/pub-v2.0.0-blue) ![GitHub](https://img.shields.io/github/license/liodali/checkbox_grouped)

    * grouped (checkbox/radioButton)
    * customisable grouped checkbox
    * grouped chips
    * grouped switch
    * recuperate the actual selection
    * make multiple selection
    * dialogGroupedCheckbox
    * list of groupedCheckbox
    * list of customisable groupedCheckbox
    * select/deselect items pragrammatically

## Getting Started
<img src="https://github.com/liodali/checkbox_grouped/blob/master/exampleCheckbox.gif?raw=true" alt="checkboxGrouped examples"><br>


## Installing

Add the following to your `pubspec.yaml` file:

    dependencies:
		checkbox_grouped: ^2.0.0



## Simple Usage
#### Creating a basic `SimpleGroupedCheckbox`

```dart
    SimpleGroupedCheckbox<int>(
            controller: controller,
            itemsTitle: ["1" ,"2","4","5"],
            values: [1,2,4,5],
            groupStyle: GroupStyle(
                  activeColor: Colors.red,
                  itemTitleStyle: TextStyle(
                  fontSize: 13
                 )
               ),
            checkFirstElement: false,
    )
```
### Declare GroupController to get selection and enabled/disabled Items

```dart
GroupController controller = GroupController();
```

####  `GroupController`
|   Properties          |  Description |
|-----------------------|--------------|
|`isMultipleSelection`  | (bool) enable multiple selection  in grouped checkbox (default:false).  |
|`initSelectedItem`     | (List) A Initialize list of values that will be selected in grouped.   |


### Get current selection

```dart
final selectedItems = controller.selectedItem;
```
### enable items

```dart
    controller.enableAll();
```
### disable all items

```dart
    controller.disableAll();
```

### enabled items

```dart
controller.enabledItemsByValues(List<T> values); 
```
----------
```dart
controller.enabledItemsByTitles(List<String> itemsTitles); 
```

### disable item

```dart
controller.disabledItemsByValues(List<T> values);
```

```dart
controller.disabledItemsByTitles(List<String> items)
```

### listener to changed values
> listener will be removed automatically when widget call dispose

```dart
controller.listen((v) {
      print("$v");
    });
```
### select/selectValues
> select one item in single selection or multiple selection

> selectValues work only in group when isMultiSelection is true

```dart
controller.select(value);
or
controller.selectValues([value]);
```

### selectAll/selectValues
> select all or some items in multiple selection group
```dart
controller.selectAll();
or
controller.selectValues([values]);
```
### deselectAll/deselectValues
> deselect all or some items in multiple selection group
```dart
controller.deselectAll();
or
controller.deselectValues([values]);
```


#### NOTICE

* those method `disabledItems` and `enabledItems` has been removed
* if you are using complex object in values , you need to implement operator == and hashcode

####  `SimpleGroupedCheckbox`
|   Properties          |  Description |
|-----------------------|--------------|
|`controller`           | (required)  Group Controller to recuperate selectionItems and disable or enableItems.  |
|`activeColor`          | (deprecated) The color to use when a CheckBox is checked.  |
|`itemsTitle`           | (required) A list of strings that describes each checkbox button. Each label must be distinct.   |
|`itemsSubTitle`        | A list of strings that describes second Text.   |
|`onItemSelected`       | Callback fire when the user make  selection    |
|`disableItems`         | Specifies which item should be disabled. The strings passed to this must match the Titles  |
|`values`               | (required) Values contains in each element.   |
|`checkFirstElement`    | `make first element in list checked`.  |
|`isExpandableTitle`    | `enable group checkbox to be expandable `.  |
|`groupTitle`           | `Text title for group checkbox`.  |
|`groupTitleStyle`      | (deprecated) `TextStyle of title for group checkbox`.  |
|`helperGroupTitle`     | `(bool) hide/show checkbox in title to help all selection or de-selection,use it when you want to disable checkbox in groupTitle default:true `.  |
|`groupTitleAlignment`  | `(Alignment) alignment of group title in group checkbox`.  |
|`groupStyle`           | `(GroupStyle) the style that should be applied on GroupedTitle,ItemTile,SubTitle`.  |

## Customisable Checkbox Grouped

#### Creating a basic `CustomGroupedCheckbox`

```dart
    CustomGroupedCheckbox<int>(
            controller:customController,
            groupTitle: "Custom GroupedCheckbox",
            itemBuilder: (ctx,index,value,isDisabled){
            return Card(
                child: Row(
                    children: <Widget>[
                        Expanded(
                            child: Align(
                                alignment: Alignment.centerLeft,
                                child: Text("$index"),
                            ),
                        ),
                        Opacity(
                              opacity: v?1:0,
                              child: Icon(Icons.check,
                                color: Colors.green,
                                size: 24,
                              ),
                         ),
                      ],
                   ),
                );
            },
            itemCount: 10,
            values: List<int>.generate(10, (i)=>i),
        )
```

### Declare CustomGroupController to get selection ,enabled/disable items

```dart
 CustomGroupController controller = CustomGroupController();
```

### Get current selection

```dart
final selectedItem = controller.selectedItem;
```

### enable items

```dart
    controller.enabledItems([value]);
```

### disable items

```dart
    controller.disabledItems([value]);

```
### listen to changed values

```dart
controller.listen((value){
 print(value);
})
```

####  `CustomGroupedCheckbox`
|   Properties          |  Description |
|-----------------------|--------------|
|`controller`           | (required) Custom Group Controller to recuperate selectionItems and disable or enableItems.  |
|`groupTitle`           | `widget title for group checkbox`.              |
|`itemBuilder`          | (required) `Called to build children.`          |
|`values`               | (required) `Values contains in each element.`   |
|`disableItems`         | (List) `values that will be disabled`   |
|`itemExtent`           | ` The extent the children are forced to have in the main axis`  |
|`isScroll`             | (bool) ` To make list of item scrollable `  |


####  `CustomGroupedCheckbox.grid`
|   Properties          |  Description |
|-----------------------|--------------|
|`controller`           | (required) Custom Group Controller to recuperate selectionItems and disable or enableItems.  |
|`groupTitle`           | `widget title for group checkbox`.              |
|`itemBuilder`          | (required) `Called to build children.`          |
|`values`               | (required) `Values contains in each element.`   |
|`isScroll`             | (bool) ` To indicate that the  list of children scrollable `  |
|`gridDelegate`         | (SliverGridDelegate) `a delegate that controls the layout of the children`   | 


## Chip grouped Usage

#### Creating a basic `SimpleGroupedChips`

```dart
SimpleGroupedChips<int>(
                    controller: controller,
                    values: [1,2,3,4,5,6,7],
                    itemTitle: ["1" ,"2","4","5","6","7"],
                    chipGroupStyle: ChipGroupStyle.minimize(
                         backgroundColorItem: Colors.red[400],
                         itemTitleStyle: TextStyle(
                         fontSize: 14,
                      ),
                    ),
                  )
```
### Declare GroupController to get selection and enabled/disabled Items

```dart
GroupController controller = GroupController();
```

####  `GroupController`
|   Properties          |  Description |
|-----------------------|--------------|
|`isMultipleSelection`  | (bool) enable multiple selection  in grouped checkbox (default:false).  |
|`initSelectedItem`     | (List) A Initialize list of values that will be selected in grouped.   |
|`chipGroupStyle`       | (ChipGroupStyle) A Initialize list of values that will be selected in grouped.   |


### Get current selection

```dart
final selectedItems = controller.selectedItem;
```
### enabled items

```dart
controller.enabledItemsByValues(List<T> values); 
```
----------
```dart
controller.enabledItemsByTitles(List<String> itemsTitles); 
```

### disable item

```dart
controller.disabledItemsByValues(List<T> values);
```

```dart
controller.disabledItemsByTitles(List<String> items)
```

####  `SimpleGroupedChip`
|   Properties          |  Description |
|-----------------------|--------------|
|`controller`           | (required) `controller to recuperate selectionItems and disable or enableItems.`  |
|`itemsTitle`           | (required) `A list of strings that describes each chip button. Each label must be distinct.`   |
|`disabledItems`        | `Specifies which item should be disabled. The strings passed to this must match the Titles`  |
|`values`               | (required) Values contains in each element.   |
|`onItemSelected`       | `Callback when users make  selection  or deselection`  |
|`isScrolling`          |`enable horizontal scrolling`.  |
|`chipGroupStyle`       | (ChipGroupStyle) `the style that uses to customize  item chip `  |

## Switch grouped Usage

#### Creating a basic `SimpleGroupedSwitch`

```dart
SimpleGroupedSwitch<int>(
                    controller: controller,
                    values: [1,2,4,5],
                    itemsTitle: ["1 " ,"2 ","4 ","5 ","6","7"],
                    groupStyle: SwitchGroupStyle(
                        itemTitleStyle: TextStyle(
                          fontSize: 16,
                          color: Colors.blue,
                       ),
                     activeColor: Colors.red,
                  ),  
             )
```
### Declare GroupController to get selection and enabled/disabled Items

```dart
GroupController controller = GroupController();
```

####  `GroupController`
|   Properties          |  Description |
|-----------------------|--------------|
|`isMultipleSelection`  | (bool) enable multiple selection  in grouped checkbox (default:false).  |
|`initSelectedItem`     | (List) A Initialize list of values that will be selected in grouped.   |


### Get current selection

```dart
final selectedItems = controller.selectedItem;
```
### enabled items

```dart
controller.enabledItemsByValues(List<T> values); 
```
----------
```dart
controller.enabledItemsByTitles(List<String> itemsTitles); 
```

### disable item

```dart
controller.disabledItemsByValues(List<T> values);
```

```dart
controller.disabledItemsByTitles(List<String> items)
```

####  `SimpleGroupedSwitch`
|   Properties          |  Description |
|-----------------------|--------------|
|`controller`           | (required) Group Controller to recuperate selectionItems and disable or enableItems.  |
|`itemsTitle`           | (required) A list of strings that describes each chip button. Each label must be distinct.   |
|`values`               | (required) Values contains in each element.   |
|`disableItems`         | Specifies which item should be disabled. The value passed to this must match the values list |
|`onItemSelected`       | Call when users make  selection  or deselection  |
|`groupStyle`           | (SwitchGroupStyle) the style that will customize text,switch  |


## showDialogGroupedCheckbox

> display  groupedCheckbox inside dialog
> return values selected

#### Creating a basic `showDialogGroupedCheckbox`

```dart
showDialogGroupedCheckbox(
                        context: context,
                        cancelDialogText: "cancel",
                        isMultiSelection: true,
                        itemsTitle: List.generate(15, (index) => "$index"),
                        submitDialogText: "select",
                        dialogTitle:Text("example dialog") ,
                        values: List.generate(15, (index) => index)
                      )
```


####  `showDialogGroupedCheckbox`
|   Properties                  |  Description |
|-------------------------------|--------------|
|`dialogTitle`                  | `(required) Text Widget that describe Title of dialog`.  |
|`itemsTitle`                   | `(required) Values contains in each element`.   |
|`values`                       | `(required) list of values`. |
|`initialSelectedValues`        | `list of initial values that you want to be selected`.    |
|`isDismissible`                |`enable multiple selection`.  |
|`cancelDialogText`             |`(string) label for cancelButton`.  |
|`submitDialogText`             |`(string) label for submitButton`.  |
|`isMultiSelection`             |`enable multiple selection`.  |

## ListGroupedCheckbox

> display  list of groupedCheckbox
> return all values selected
>
### Declare ListGroupController to get all item selected or get item selected by group index

```dart
ListGroupController controller = ListGroupController();
```

####  `ListGroupController`
|   Properties                  |  Description |
|------------------------------ |--------------|
|`isMultipleSelectionPerGroup`  | (List<bool>)  enable multiple selection  in each grouped checkbox. |
|`initSelectedValues`           | (List) A Initialize list of values on each group of checkbox that will be selected in group.   |


### Get current selection
* get all selection

```dart
final selectedItems = controller.allSelectedItems;
```
* get all selection by group

```dart
final selectedItems = controller.selectedItemsByGroupIndex(indexGroup);
```
### enable items

```dart
    controller.enableAll(int indexGroup);
```
### disable all items

```dart
    controller.disableAll(int indexGroup);
```
### enabled items

```dart
controller.enabledItemsByValues(int indexGroup,List<T> values); 
```
----------
```dart
controller.enabledItemsByTitles(int indexGroup,List<String> itemsTitles); 
```

### disable item

```dart
controller.disabledItemsByValues(int indexGroup,List<T> values);
```

```dart
controller.disabledItemsByTitles(int indexGroup,List<String> items)
```



#### Creating a basic `ListGroupedCheckbox`


```dart
                ListGroupedCheckbox(
                        controller: listController,
                        groupTitles: List.generate(3, (index) => "groupe $index"),
                        values: List.generate(
                          3,
                          (i) =>
                              List.generate(5, (j) => "${(i + Random().nextInt(100)) * j}"),
                        ),
                        titles: List.generate(
                          3,
                          (i) => List.generate(5, (j) => "Title:$i-$j"),
                        ),
                        mapItemGroupedType: {
                          1: GroupedType.Chips,
                        },
                      )
```


####  `ListGroupedCheckbox`
|   Properties                        |  Description |
|-------------------------------------|--------------------------------------------------------------------------------------------------|
|`controller`                         | (required) manage the ListGroupedCheckbox.  |
|`groupTitles`                        | (required)Text title for group checkbox in each groupedCheckbox.  |
|`titles`                             | (required) A list of list of strings that describes each checkbox button. Each label must be distinct in groupedCheckbox.   |
|`values`                             | (required) Values contains in each element in each groupedCheckbox.   |
|`subTitles`                          | A list of list strings that describes second Text in each groupedChckbox.   |
|`isScrollable`                       | (bool) make the parent widget scrollable or disable it (default:true).   |
|`onSelectedGroupChanged`             | CallBack to get all selected items when users  make select new items or deselect items  |
|`disabledValues`                     | A nested list of string ,specifies which item should be disabled in each groupedCheckbox. The strings passed to this must match the Titles  |
|`titleGroupedAlignment`              | (Alignment) align text title of each group of checkbox  |
|`titleGroupedTextStyle`              | (TextStyle) style text title of each group   |
|`mapItemGroupedType`                 | (Map) define type of each group can be chips/switch/(checkbox or radio)  |

------------------

##ListCustomGroupedCheckbox
> display  list of groupedCheckbox
> return all values selected
>
### Declare ListGroupController to get all item selected or get item selected by group index

```dart
ListGroupController controller = ListGroupController();
```

####  `ListCustomGroupController`
|   Properties                  |  Description |
|------------------------------ |--------------|
|`isMultipleSelectionPerGroup`  | (List<bool>)  enable multiple selection  in each grouped checkbox. |
|`initSelectedValuesByGroup`    | (List) A Initialize list of values on each group of checkbox that will be selected in group.   |


### Get current selection
* get all selection
1) list
```dart
final selectedItems = controller.allSelectedItems;
```
2) map
```dart
final selectedItems = controller.mapSelectedItems;
```
* get all selection by group

```dart
final selectedItems = controller.selectedItemsByGroupIndex(indexGroup);
```

####  `ListCustomGroupedCheckbox`
|   Properties                        |  Description |
|-------------------------------------|--------------------------------------------------------------------------------------------------|
|`controller`                         | (required) manage the ListGroupedCheckbox.  |
|`children`                           | (required) list of CustomIndexedWidgetBuilder to return widget of items for each group,should be non null widgets  |
|`groupTitles`                        | A list of String for group checkbox in each CustomGrouped.  |
|`groupTitlesWidget`                  | A list of list of widgets that describes custom Title for each group.   |
|`listValuesByGroup`                  | (required) Values contains in each element in each CustomGroupedCheckbox.   |
|`isScrollable`                       | (bool) make the parent widget scrollable or disable it (default:true).   |
|`onSelectedGroupChanged`             | CallBack to get all selected items when one of item hit it make select new items or deselect items  |
|`titleGroupedAlignment`              | (Alignment) align text title of each group  |
|`titleGroupedTextStyle`              | (TextStyle) style text for title of each group   |


### basic example 
```dart
final controller =
ListCustomGroupController(isMultipleSelectionPerGroup: [true, false]);
final data = [
  List<_User>.generate(
    10,
        (i) => _User(
      name: faker.person.name(),
      email: faker.internet.email(),
    ),
  ),
  List<String>.generate(
    10,
        (i) => faker.person.name(),
  )
]
ListCustomGroupedCheckbox(
      controller: controller,
      onSelectedGroupChanged: (values) {
        print(values);
      },
      isScrollable: true,
      titleGroupedAlignment: Alignment.centerLeft,
      groupTitles: ["Users", "Names"],
      children: [
        CustomIndexedWidgetBuilder(
            itemBuilder: (ctx, index, selected, isDisabled) {
          return Card(
            margin: EdgeInsets.only(
              top: 5.0,
              bottom: 5.0,
            ),
            child: ListTile(
              tileColor: selected ? Colors.green[300] : Colors.white,
              title: Text(
                (data[0][index] as _User).name,
              ),
              subtitle: Text(
                (data[0][index] as _User).email,
              ),
              trailing: Opacity(
                opacity: selected ? 1 : 0,
                child: Icon(
                  Icons.check,
                  color: Colors.white,
                  size: 24,
                ),
              ),
            ),
          );
        }),
        CustomGridIndexedWidgetBuilder(
            itemBuilder: (ctx, index, selected, isDisabled) {
              return Card(
                color: selected ? Colors.green[300] : Colors.white,
                shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(
                    12.0,
                  ),
                ),
                margin: EdgeInsets.only(
                  top: 5.0,
                  bottom: 5.0,
                ),
                child: Stack(
                  children: [
                    Center(
                      child: Text(
                        data[1][index],
                        maxLines: 3,
                        overflow: TextOverflow.fade,
                      ),
                    ),
                    Positioned(
                      top: 5,
                      right: 8,
                      child: Opacity(
                        opacity: selected ? 1 : 0,
                        child: Icon(
                          Icons.check,
                          color: Colors.white,
                          size: 24,
                        ),
                      ),
                    ),
                  ],
                ),
              );
            },
            gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                      crossAxisCount: 3, 
                      crossAxisSpacing: 5.0,

CHANGELOG.md
## [2.0.0] - improve some internal api,fix bug #62
* create class `BaseSimpleGrouped`,`BaeCustomGrouped` to make sure our widget will keep state reloading
* add disableItems to `CustomGroupedCheckbox`
* fix some bugs
## [1.9.0] - fix bug #53
* migrate to latest flutter version and support only null safety
* fix bug in base stateGroup
## [1.8.2] - fix bug #67
* ignore pre-selection if not exist in list of values
* replace listView.builder in lisg_group by SingleChildScrollView
## [1.8.1] - fix bug 
* groupTitlesWidget not taking in consideration in list custom group
## [1.8.0] - 
* add new method in custom_group to clear all selection items
## [1.7.1] - fix bug
* fix subItem style of the groupStyle doesn't applied correctly (#59)
## [1.7.0] - add new widget ListCustomGroupedCheckbox 
*  * create ListCustomGroupedCheckbox to manage multiple custom grouped in list  
*  * create ListCustomController to get selected item 
*  * remove deprecated attributes  
## [1.6.1+1] - fix  readme
## [1.6.1] - fix  bug [#55]
* add new attribute `isScrollable` to make widget scrollable or not (default:true)
## [1.6.0] - add styles for SimpleGroupedCheckbox,SimpleGroupedChips,SimpleGroupedSwitch 
* create new methods in GroupController :`select`,`selectValues`,`selectAll`,`deselectValues` and `deselectAll`
## [1.5.0] - add styles for SimpleGroupedCheckbox,SimpleGroupedChips,SimpleGroupedSwitch 
*  create new generic `GroupedStyle` to manage style of simple_grouped_checkbox 
*  deprecate `backgroundColorItem`,`selectedColorItem`,`textColor`,`selectedTextColor`,`selectedIcon`,`disabledColor` 
*  add `chipGroupStyle` in SimpleGroupedChips 
*  add `SwitchGroupStyle` style for SimpleGroupedSwitch 
*  add `SwitchGroupStyle` style for SimpleGroupedSwitch 
## [1.4.2] - fix bugs 
*  fix bugs in helper expandable simple grouped 
## [1.4.1] - fix bugs 
*  fix bugs in simple grouped 
*  fix bugs in init selected of custom grouped 
## [1.4.0] - add new method in `BaseController` and `BaseListController`
* add `enabledAll` and `disabledAll` in `BaseController` and `BaseListController`
* add unit test  for `enabledAll` and `disabledAll`
* fix bug
## [1.3.0] - show different type of group checkbox in list grouped
* show  multiple type (chips,switch,(radio/checkbox)) inside list of grouped
* fix bugs
## [1.2.0-nullsafety.0] - add listener to groupController
* create new method `listen` to get selected value from controller out side of widget
## [1.1.0-nullsafety.1] - change gif example
## [1.1.0-nullsafety.0] - grid custom grouped checkbox
* add grid custom grouped checkbox
* add `listen` method to add listener in changed data in real time
* add `isScroll` attribute in custom grouped checkbox
* fix bugs
## [1.0.1-nullsafety.0] - fix bugs,refactoring code 
*  fix bugs in chips grouped 
*  ameliorate null safety generated code 
*  fix bugs unit test 
## [1.0.0-nullsafety.1] - remove unnecessary files
* change gis example 
* remove asset of flare 
## [1.0.0-nullsafety.0] -nullsafety migration
* remove dependencies that doesn't support null safety
* remove circulaire checkbox for now until will reimplemented 
* nullsafety migration
## [0.8.1] -improve readme 
* change example 
* fix bug
## [0.8.0+1] -improve readme 
*  add docs,fix readme 
## [0.8.0] -improve readability of the code and  remove working with globalKey 
*  add GroupController to control simple grouped checkbox/switch/chips 
*  add CustomGroupController to control custom grouped checkbox 
*  add ListGroupController to control list grouped checkbox 
*  add enable/disable items in all groups checkbox 
*  modify example with new changes 
*  add and update widget unit tests 
## [0.7.0+1] - add groupTitleAlignment,helperGroupTitle checkbox 
*  add helperGroupTitle to show or hid checkbox in groupTile 
*  add groupTitleAlignment to change alignment of groupTitle  of 2 new method disableItemsByTitles 
## [0.7.0] - add new method to disable and enabled checkbox 
*  deprecation of the method disabledItems and enabledItems
*  creation  of 2 new method disabledItemsByTitles and disabledItemsByValues to disable checkbox items
*  creation  of 2 new method enabledItemsByTitles and enabledItemsByValues to enable checkbox items
## [0.6.3] - fix null exception 
*  fix selection method in custom_grouped_checkbox
* fix url gif in readme
## [0.6.2] - add gif to readme 
*  add new gif to show basics examples of the library 
## [0.6.2] - pre-selection in single choice 
*  add possibility to pre-selection any element in single chice 
## [0.6.1+1] - fix color of groupTitle in dark mode 
*  remove color in textStyle of groupTitle 
*  add textStyle to modify Style of groupTitle 
## [0.6.1] - add callback to collect data from list of grouped widget 
*  create typedef onGroupChanged and add it to widget to collect data inside widget 
## [0.6.0+1] - fix bug 
*  fix bug of unchanged state of checkbox in titleGrouped 
*  remove widget in simpleGroupedCheckbox (example 1) 
## [0.6.0] - add new widget:ListGroupedCheckbox 
*  ListGroupedCheckbox to manage list of groupedCheckbox,retrieve selected data in one call 
*  add example for ListGroupedCheckbox 
*  add test widget for ListGroupedCheckbox 
## [0.5.2] - optimize simple grouped switch
*  use valueNotifier to manage selection in simple grouped switch 
*  make physics of ListView non Scrollable 
*  update kotlin version in android project of the example  
## [0.5.1] - optimize simple grouped chips 
*  use valueNotifier to manage selection in simple grouped chips 
*  change main example  
*  fix bug in test simple grouped chips 
*  add test in multiple choice for simple grouped chips 
## [0.5.0+1] - fix bug in simple grouped chips 
*  fix bug related to assert 
## [0.5.0] - improve performance and create testsWidget 
*  improve performance of simple groupedCheckbox:use valueNotifier,remove setState 
*  fix and recreate test widget of simple groupedCheckbox 
## [0.4.5+7] - fix issues 
*  update dependencies 

example/lib/main.dart
import 'package:checkbox_grouped/checkbox_grouped.dart';
import 'package:example/custom_grid_grouped_example.dart';
import 'package:flutter/material.dart';

import 'custom_grouped_example.dart';
import 'grid_of_grouped_checkbox.dart';
import 'list_custom_group.dart';
import 'list_of_grouped.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return _MyAppState();
  }
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        useMaterial3: true,
        primaryColor: Colors.blue,
      ),
      home: MainExample(),
    );
  }
}

class MainExample extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return _MainExampleState();
  }
}

class _MainExampleState extends State<MainExample>
    with TickerProviderStateMixin {
  late TabController tabController;
  ValueNotifier<int> current = ValueNotifier(0);
  final customController = CustomGroupController(
    isMultipleSelection: false,
    initSelectedItem: "basics",
  );
  final List<String> drawerItems = [
    "basics",
    "custom",
    "grid custom",
    "dialog",
    "list of group",
    "more example",
    "list of custom group"
  ];

  void tabChanged() {
    current.value = tabController.index;
  }

  @override
  void initState() {
    super.initState();
    tabController = TabController(
      initialIndex: current.value,
      length: drawerItems.length,
      vsync: this,
    );
    tabController.addListener(tabChanged);
    customController.listen((value) {
      final index = drawerItems.indexOf(value);
      if (index != -1) {
        tabController.index = index;
        Navigator.pop(context);
      }
    });
  }

  @override
  void dispose() {
    tabController.removeListener(tabChanged);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      drawer: Drawer(
        child: ValueListenableBuilder<int>(
          valueListenable: current,
          builder: (ctx, value, _) {
            return Column(
              children: [
                SizedBox(
                  height: 56,
                ),
                Expanded(
                  child: CustomGroupedCheckbox<String>(
                    controller: customController,
                    itemBuilder: (ctx, index, isSelected, isDisabled) {
                      return ListTile(
                        title: Text(
                          drawerItems[index],
                          style: TextStyle(
                            color: value == index ? Colors.blue : null,
                          ),
                        ),
                      );
                    },
                    itemExtent: 64,
                    values: drawerItems,
                  ),
                ),
              ],
            );
          },
        ),
      ),
      appBar: AppBar(
        title: Text("examples"),
      ),
      body: TabBarView(
        controller: tabController,
        physics: NeverScrollableScrollPhysics(),
        children: <Widget>[
          _SimpleGrouped(),
          CustomGroupedExample(),
          CustomGridGroupedExample(),
          _DialogExample(),
          ListOfGrouped(),
          GridOfListGroupedCheckbox(),
          ListCustomGroup(),
        ],
      ),
    );
  }
}

class _SimpleGrouped extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => _SimpleGroupedState();
}

class _SimpleGroupedState extends State<_SimpleGrouped> {
  GroupController controller = GroupController(initSelectedItem: [2]);
  GroupController switchController = GroupController();
  GroupController chipsController = GroupController(isMultipleSelection: true);
  GroupController multipleCheckController = GroupController(
    isMultipleSelection: true,
    initSelectedItem: List.generate(10, (index) => index),
  );

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      controller: ScrollController(),
      child: Column(
        mainAxisSize: MainAxisSize.max,
        children: <Widget>[
          SimpleGroupedCheckbox<int>(
            controller: controller,
            //groupTitle:"Basic",
            onItemSelected: (data) {
              print(data);
              if (data == 1) {
                controller.disabledItemsByTitles(["5"]);
              } else if (data == 4) {
                controller.enabledItemsByTitles(["5", "2"]);
                controller.disabledItemsByTitles(["1"]);
              } else if (data == 2) {
                controller.enabledItemsByTitles(["1"]);
              }
            },
            disableItems: [5, 4],
            itemsTitle: ["1", "2", "4", "5", "6"],
            values: [1, 2, 4, 5, 6],
            groupStyle: GroupStyle(
              activeColor: Colors.red,
              itemTitleStyle: TextStyle(
                fontSize: 13,
              ),
            ),
          ),
          SimpleGroupedCheckbox<int>(
            controller: multipleCheckController,
            itemsTitle: List.generate(10, (index) => "$index"),
            values: List.generate(10, (index) => index),
            groupStyle: GroupStyle(
              activeColor: Colors.green,
              groupTitleStyle: TextStyle(
                color: Colors.orange,
              ),
            ),
            groupTitle: "expanded multiple checkbox selection",
            helperGroupTitle: true,
            onItemSelected: (data) {
              print(data);
            },
            isExpandableTitle: true,
          ),
          Divider(),
          SimpleGroupedChips<int>(
            controller: chipsController,
            values: List.generate(7, (index) => index),
            itemsTitle: List.generate(7, (index) => "chip_text_$index"),
            chipGroupStyle: ChipGroupStyle.minimize(
              isScrolling: true,
              backgroundColorItem: Colors.grey[400],
              selectedTextColor: Colors.amber,
              itemTitleStyle: TextStyle(
                fontSize: 14,
              ),
              shape: RoundedRectangleBorder(
                borderRadius: BorderRadius.circular(12),
                side: BorderSide.none,
              ),
              checkedShape: RoundedRectangleBorder(
                borderRadius: BorderRadius.circular(12),
                side: BorderSide(
                  width: 2,
                  color: Colors.grey,
                ),
              ),
              selectedIcon: Icons.check,
              direction: ChipsDirection.horizontal,
            ),
            onItemSelected: (values) {
              print(values);
            },
          ),
          Divider(),
          Text("grouped switch"),
          SimpleGroupedSwitch<int>(
            controller: switchController,
            itemsTitle: List.generate(7, (index) => "${index}"),
            values: List.generate(7, (index) => index),
            disableItems: [3],
            groupStyle: SwitchGroupStyle(
              itemTitleStyle: TextStyle(
                fontSize: 16,
                color: Colors.blue,
              ),
              activeColor: Colors.red,
            ),
            onItemSelected: (values) {
              print(values);
            },
          ),
        ],
      ),
    );
  }
}

class _DialogExample extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Column(
      children: <Widget>[
        TextButton(
          onPressed: () async {
            var values = await showDialogGroupedCheckbox(
                context: context,
                cancelDialogText: "cancel",
                isMultiSelection: true,
                itemsTitle: List.generate(15, (index) => "$index"),
                submitDialogText: "select",
                dialogTitle: Text("example dialog"),
                values: List.generate(15, (index) => index));
            if (values != null) {
              print(values);
            }
          },
          child: Text("show dialog checkbox grouped"),
        ),
      ],
    );
  }
}
